programs/server/Server.cpp

```diff
@@ -594,6 +594,9 @@ int Server::main(const std::vector<std::string> & /*args*/)
     if (!access_control_local_path.empty())
         global_context->getAccessControlManager().setLocalDirectory(access_control_local_path);
 
+    /// Sets external authenticators config (LDAP).
+    global_context->setExternalAuthenticatorsConfig(config());
```

Comment:
Isn't it better to move that call into `auto main_config_reloader = std::make_unique<ConfigReloader>(` part to allow rereading the configuration after its change without restarting ClickHouse?

Reply:
Re-reading the configuration without restarting ClickHouse would be very nice and would speed up all the tests. We currently need to restart the server whenever we change LDAP server descriptions or users in the configuration files.

Reply:
Pushed a modification, that should reload `ExternalAuthenticators` config too. @vzakaznikov could you please include tests to verify that?

Reply:
This change causes a crash when LDAP servers to the configuration file and server is restarted.
```
2020.07.10 23:06:02.060267 [ 47 ] {} <Information> TCPHandler: Processed in 0.025695653 sec.
2020.07.10 23:06:02.060348 [ 47 ] {} <Information> TCPHandler: Done processing connection.
2020.07.10 23:06:02.073618 [ 47 ] {} <Trace> TCPHandlerFactory: TCP Request. Address: 127.0.0.1:45272
2020.07.10 23:06:02.073677 [ 47 ] {} <Debug> TCPHandler: Connected ClickHouse client version 20.6.0, revision: 54436, user: user1.
2020.07.10 23:06:02.073716 [ 47 ] {} <Trace> ContextAccess (user1): Settings: readonly=0, allow_ddl=true, allow_introspection_functions=false
2020.07.10 23:06:02.073729 [ 47 ] {} <Trace> ContextAccess (user1): List of all grants: GRANT SHOW, SELECT, INSERT, ALTER, CREATE, DROP, TRUNCATE, OPTIMIZE, KILL QUERY, SYSTEM, dictGet, SOURCES ON *.*
2020.07.10 23:06:02.073892 [ 8 ] {} <Trace> BaseDaemon: Received signal 11
2020.07.10 23:06:02.073998 [ 70 ] {} <Fatal> BaseDaemon: ########################################
2020.07.10 23:06:02.074021 [ 70 ] {} <Fatal> BaseDaemon: (version 20.6.1.1, build id: AC3582F40E227CA7DE9C1AB5BD8ADCEEC44DAA45) (from thread 47) (no query) Received signal Segmentation fault (11)
2020.07.10 23:06:02.074040 [ 70 ] {} <Fatal> BaseDaemon: Address: 0x10 Access: read. Address not mapped to object.
2020.07.10 23:06:02.074050 [ 70 ] {} <Fatal> BaseDaemon: Stack trace: 0x7f581f246084 0x560943e 0xdaba6a9 0x98368bf 0xa3b1c27 0x97fb83c 0x9a56068 0xa130366 0xa133d39 0xa1352bc 0xcaf42e7 0xcaf46eb 0xcc335be 0xcc2ff2c 0x7f581f243669 0x7f581f15a323
2020.07.10 23:06:02.074087 [ 70 ] {} <Fatal> BaseDaemon: 3. pthread_mutex_lock @ 0xc084 in /usr/lib/x86_64-linux-gnu/libpthread-2.30.so
2020.07.10 23:06:02.074208 [ 70 ] {} <Fatal> BaseDaemon: 4. /home/vzakaznikov/github/ClickHouse-traceon/build/../contrib/libcxx/include/atomic:746: pthread_mutex_lock @ 0x560943e in /usr/bin/clickhouse
2020.07.10 23:06:02.074712 [ 70 ] {} <Fatal> BaseDaemon: 5. /home/vzakaznikov/github/ClickHouse-traceon/build/../contrib/libcxx/src/mutex.cpp:34: std::__1::mutex::lock() @ 0xdaba6a9 in /usr/bin/clickhouse
2020.07.10 23:06:02.074956 [ 70 ] {} <Fatal> BaseDaemon: 6. /home/vzakaznikov/github/ClickHouse-traceon/build/../contrib/libcxx/include/map:1380: DB::ExternalAuthenticators::getLDAPServerParams(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) const @ 0x98368bf in /usr/bin/clickhouse
2020.07.10 23:06:02.075346 [ 70 ] {} <Fatal> BaseDaemon: 7. /home/vzakaznikov/github/ClickHouse-traceon/build/../src/Access/Authentication.cpp:86: DB::Authentication::isCorrectPassword(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, DB::ExternalAuthenticators const&) const @ 0xa3b1c27 in /usr/bin/clickhouse
2020.07.10 23:06:02.075606 [ 70 ] {} <Fatal> BaseDaemon: 8. /home/vzakaznikov/github/ClickHouse-traceon/build/../src/Access/ContextAccess.cpp:214: DB::ContextAccess::isCorrectPassword(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) const @ 0x97fb83c in /usr/bin/clickhouse
2020.07.10 23:06:02.076214 [ 70 ] {} <Fatal> BaseDaemon: 9. /home/vzakaznikov/github/ClickHouse-traceon/build/../src/Interpreters/Context.cpp:681: DB::Context::setUser(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, Poco::Net::SocketAddress const&) @ 0x9a56068 in /usr/bin/clickhouse
2020.07.10 23:06:02.076704 [ 70 ] {} <Fatal> BaseDaemon: 10. /home/vzakaznikov/github/ClickHouse-traceon/build/../src/Server/TCPHandler.cpp:745: DB::TCPHandler::receiveHello() @ 0xa130366 in /usr/bin/clickhouse
2020.07.10 23:06:02.077358 [ 70 ] {} <Fatal> BaseDaemon: 11. /home/vzakaznikov/github/ClickHouse-traceon/build/../contrib/libcxx/include/string:1426: DB::TCPHandler::runImpl() @ 0xa133d39 in /usr/bin/clickhouse
2020.07.10 23:06:02.078082 [ 70 ] {} <Fatal> BaseDaemon: 12. /home/vzakaznikov/github/ClickHouse-traceon/build/../src/Server/TCPHandler.cpp:1203: DB::TCPHandler::run() @ 0xa1352bc in /usr/bin/clickhouse
2020.07.10 23:06:02.078392 [ 70 ] {} <Fatal> BaseDaemon: 13. /home/vzakaznikov/github/ClickHouse-traceon/build/../contrib/poco/Net/src/TCPServerConnection.cpp:57: Poco::Net::TCPServerConnection::start() @ 0xcaf42e7 in /usr/bin/clickhouse
2020.07.10 23:06:02.078712 [ 70 ] {} <Fatal> BaseDaemon: 14. /home/vzakaznikov/github/ClickHouse-traceon/build/../contrib/libcxx/include/atomic:856: Poco::Net::TCPServerDispatcher::run() @ 0xcaf46eb in /usr/bin/clickhouse
2020.07.10 23:06:02.079049 [ 70 ] {} <Fatal> BaseDaemon: 15. /home/vzakaznikov/github/ClickHouse-traceon/build/../contrib/poco/Foundation/include/Poco/Mutex_POSIX.h:59: Poco::PooledThread::run() @ 0xcc335be in /usr/bin/clickhouse
2020.07.10 23:06:02.079352 [ 70 ] {} <Fatal> BaseDaemon: 16. /home/vzakaznikov/github/ClickHouse-traceon/build/../contrib/poco/Foundation/include/Poco/AutoPtr.h:223: Poco::ThreadImpl::runnableEntry(void*) @ 0xcc2ff2c in /usr/bin/clickhouse
2020.07.10 23:06:02.079364 [ 70 ] {} <Fatal> BaseDaemon: 17. start_thread @ 0x9669 in /usr/lib/x86_64-linux-gnu/libpthread-2.30.so
2020.07.10 23:06:02.079375 [ 70 ] {} <Fatal> BaseDaemon: 18. __clone @ 0x122323 in /usr/lib/x86_64-linux-gnu/libc-2.30.so
2020.07.10 23:06:02.079381 [ 70 ] {} <Information> SentryWriter: Not sending crash report
2020.07.10 23:06:03.655593 [ 55 ] {} <Debug> ConfigReloader: Loading config '/etc/clickhouse-server/config.xml'
2020.07.10 23:06:03.656675 [ 55 ] {} <Debug> ConfigReloader: Loaded config '/etc/clickhouse-server/config.xml', performing update on configuration
2020.07.10 23:06:04.708822 [ 48 ] {} <Trace> TCPHandlerFactory: TCP Request. Address: 127.0.0.1:45278
2020.07.10 23:06:04.708904 [ 48 ] {} <Debug> TCPHandler: Connected ClickHouse client version 20.6.0, revision: 54436, user: default.
2020.07.10 23:06:09.152276 [ 32 ] {} <Trace> SystemLog (system.query_log): Flushing system log, 2 entries to flush
2020.07.10 23:06:09.152284 [ 34 ] {} <Trace> SystemLog (system.query_thread_log): Flushing system log, 2 entries to flush
2020.07.10 23:06:09.152423 [ 34 ] {} <Debug> SystemLog (system.query_thread_log): Will use existing table system.query_thread_log for QueryThreadLog
2020.07.10 23:06:09.152424 [ 32 ] {} <Debug> SystemLog (system.query_log): Will use existing table system.query_log for QueryLog
2020.07.10 23:06:09.728105 [ 49 ] {} <Trace> TCPHandlerFactory: TCP Request. Address: 127.0.0.1:45286
2020.07.10 23:06:09.728194 [ 49 ] {} <Debug> TCPHandler: Connected ClickHouse client version 20.6.0, revision: 54436, user: default.
```

Reply:
Crash is fixed, but the issue still exists. Tests need to be adjusted to make sure they execute queries after the config has been parsed and applied.

Reply:
@vzakaznikov With the latest changes, now we have `DB::Exception: LDAP server 'openldap1' is not configured.` instead of `DB::Exception: External authenticators are not configured`, but the cause is the same: client attempts to connect, while config hasn't been loaded/applied yet.